Yoast/javascript
# Add Arabic and Urdu punctuation marks for word tokenization

## Summary


This PR can be summarized in the following changelog entry:

* [yoastseo] Improves word tokenization when words are followed by specific Arabic or Urdu punctuation marks.
* [Yoast SEO Free] Improves keyphrase and prominent word recognition when words in the text occur with specific Arabic or Urdu punctuation marks.
* [Yoast SEO Premium] Improves prominent word recognition when words in the text occur with specific Arabic or Urdu punctuation marks.

## Relevant technical choices:

* Unfortunately, word boundaries are defined in 4 different places in our code. This made it necessary to add the punctuation marks in 4 different spots. Since there are slight differences in how the word boundaries are defined and used, doing a thorough refactor would take quite some time and extra testing. I made the decision not to refactor now to keep this PR as small in scope as possible.
* Below, four different parts of functionality are used to test the different word boundary definitions. Of course, the word boundaries are also used in other assessments, but for the current purposes I picked one assessment to test each word boundary definition. Here is an overview how each assessment relates to each word boundary definition
  * `determineProminentWords` -> `getWords` -> `removePunctuation`
  * `matchKeywordInSubheading` -> `findKeywordFormsInstring`/`findTopicFormsInString` -> `matchTextWithArray` -> `matchTextWithWord` -> `matchTextWithTransliteration` -> `addWordBoundary`
  * `findTransitionWords` -> `matchWordInSentence`/`isWordInSenence` -> `wordBoundaries`
  * `keywordDensityAssessment` -> `keywordCount` -> `markWordsInSentences` -> `addMarkSingleWord` -> `stripWordBoundariesEnd` -> `stripWordBoundaries` 
* Based on `release/15.0` to be able to test with the transition word assessment which is made available for Arabic on that branch.

## Test instructions

This PR can be tested by following these steps:

* Set your website language to Arabic.

Test in Free or Premium:
* Test the keyword density assessment
  * Add an [Arabic text](https://generator.lorem-ipsum.info/_arabic) with ≥150 words.
  * Add the punctuation mark `؟` behind (meaning on the left of) a word, e.g. `لفرنسا؟`.
  * Set that word as the keyphrase, without the punctuation mark, e.g. `لفرنسا`.
  * Make sure the word gets counted and marked by the Keyphrase density assessment.
* Test the transition word assessment
  * Add an Arabic sentence. Make sure it contains no transition word (as indicated by the transition words assessment).
  * Add an Arabic transition word, e.g. `حاليا`.
  * The transition word assessment should now recognize and mark that sentence.
  * Add the punctuation mark `،` behind the transition word, i.e. `،حاليا`.
  * The sentence should still be recognized and marked by the transition word assessment.
* Test the keyphrase in subheading assessment
  * Add an H2 or H3 subheading to your text with an Arabic word followed by `؟`, e.g. `قط؟`.
  * Set `قط` as your keyphrase.
  * You should get a green bullet with the feedback `Keyphrase in subheading: Your H2 or H3 subheading reflects the topic of your copy. Good job!`.

Test in Premium:
* Test prominent words
  * Add the text  `قط؟ قط؟ قط؟ قط؟ قط؟ قط؟ قط؟`.
  * In your insights the word `قط` should be shown without the `؟`.

## Impact check

* This PR affects the following parts of the plugin, which may require extra testing:
  *

## UI changes

## Quality assurance


Fixes https://yoast.atlassian.net/browse/LIN-563
